File: runtimes/rust/src/pkg/conf.rs
use config::{Config, ConfigError, Environment};
use lazy_static::lazy_static;
use serde::Deserialize;

#[derive(Deserialize)]
pub struct Settings {
    pub project: String,
    pub name: String,

    pub database_url: String,
    pub redis_url: String,
    pub nats_broker_url: String,
    //otel
    pub otlp_host: Option<String>,
    pub otlp_port: Option<String>,
    pub use_telemetry: bool,
}

impl Settings {
    pub fn new() -> Result<Self, ConfigError> {
        let conf = Config::builder()
            .add_source(Environment::default())
            .build()?;
        conf.try_deserialize()
    }
}

lazy_static! {
    pub static ref settings: Settings = Settings::new().expect("improperly configured");
}

File: runtimes/go/pkg/conf.go
package pkg

import (
	"fmt"
	"sync"

	"go-simpler.org/env"
)

type Settings struct {
	Project string `env:"PROJECT"`
	Name    string `env:"NAME"`

	DatabaseUrl   string `env:"DATABASE_URL"`
	RedisUrl      string `env:"REDIS_URL"`
	RedisPassword string `env:"REDIS_PASSWORD"`
	NatsUrl       string `env:"NATS_URL"`

	OtlpHost     string `env:"OTLP_HOST"`
	OtlpPort     string `env:"OTLP_PORT"`
	UseTelemetry bool   `env:"USE_TELEMETRY"`
}

var (
	settings *Settings
	once     sync.Once
)

func LoadSettings() *Settings {
	once.Do(func() {
		settings = &Settings{}
		err := env.Load(settings, nil)
		if err != nil {
			fmt.Printf("%s", err)
		}
	})
	return settings
}
